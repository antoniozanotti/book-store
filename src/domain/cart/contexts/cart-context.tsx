"use client";

import {
  createContext,
  useContext,
  useReducer,
  ReactNode,
  useEffect,
} from "react";
import { CartType } from "../types/cart-type";
import { CartContextType } from "../types/cart-context-type";
import { useLocalStorage } from "usehooks-ts";
import { useProductsByIdsQuery } from "@/domain/product/queries/use-products-by-ids-query";
import { cartReducer } from "../reducers/cart-reducer";

const initialState: CartType = { items: [] };

const CartContext = createContext<CartContextType>({
  state: initialState,
  dispatch: () => {},
  getCartItemsQty: () => {},
  cartItemsSubtotal: () => {},
});

export function CartProvider({ children }: { children: ReactNode }) {
  const [localState, setLocalState] = useLocalStorage<CartType>(
    "cart",
    initialState
  );
  const [state, dispatch] = useReducer(cartReducer, localState);
  const cartItems = state.items;
  const responses = useProductsByIdsQuery(cartItems.map((item) => item.id));

  function getCartItemsQty(): number {
    const cartItems = state.items;
    return cartItems.reduce((acc, cur) => {
      return acc + cur.quantity;
    }, 0);
  }

  function cartItemsSubtotal(): number {
    let subTotal = 0;
    cartItems.forEach((item) => {
      responses.some((response) => {
        const product = response.data;
        if (!product) {
          return true;
        }
        if (item.id == product.id) {
          subTotal += parseInt(product.price.toString()) * item.quantity;
          return true;
        }
      });
    });
    return subTotal;
  }

  useEffect(() => {
    setLocalState(state);
  }, [state, setLocalState]);

  return (
    <CartContext.Provider
      value={{ state, dispatch, getCartItemsQty, cartItemsSubtotal }}
    >
      {children}
    </CartContext.Provider>
  );
}

export function useCartContext(): CartContextType {
  return useContext(CartContext);
}
