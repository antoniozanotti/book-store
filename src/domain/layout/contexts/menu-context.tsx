"use client";

import { createContext, useContext, useReducer, ReactNode } from "react";
import { MenuType } from "../types/menu-type";
import { MenuContextType } from "../types/menu-context-type";
import { menuReducer } from "../reducers/menu-reducer";

const initialState: MenuType = { isOpen: false };

const MenuContext = createContext<MenuContextType>({
  state: initialState,
  dispatch: () => {},
});

export function MenuProvider({ children }: { children: ReactNode }) {
  const [state, dispatch] = useReducer(menuReducer, initialState);

  return (
    <MenuContext.Provider value={{ state, dispatch }}>
      {children}
    </MenuContext.Provider>
  );
}

export function useMenuContext(): MenuContextType {
  return useContext(MenuContext);
}
