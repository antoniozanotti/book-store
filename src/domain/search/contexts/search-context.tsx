"use client";

import { createContext, useContext, useReducer, ReactNode } from "react";
import { SearchContextType } from "../types/search-context-type";
import { useSearchParams } from "next/navigation";
import { getSearchParams } from "../services/search-service";
import { searchReducer } from "../reducers/search-reducer";

const SearchContext = createContext<SearchContextType>({} as SearchContextType);

export function SearchProvider({ children }: { children: ReactNode }) {
  const searchParams = useSearchParams();
  const { keyword, categories, minPrice, maxPrice, minYear, maxYear, orderBy } =
    getSearchParams(searchParams);
  const initialStateBySearchParams = {
    keyword: keyword,
    categories: categories,
    minprice: minPrice,
    maxprice: maxPrice,
    minyear: minYear,
    maxyear: maxYear,
    orderby: orderBy,
  };
  const [state, dispatch] = useReducer(
    searchReducer,
    initialStateBySearchParams
  );

  return (
    <SearchContext.Provider value={{ state, dispatch }}>
      {children}
    </SearchContext.Provider>
  );
}

export function useSearchContext(): SearchContextType {
  return useContext(SearchContext);
}
