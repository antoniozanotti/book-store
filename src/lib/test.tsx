import React, { ReactElement, ReactNode } from "react";
import {
  render,
  RenderOptions,
  RenderResult,
  Screen,
} from "@testing-library/react";

function AllTheProviders({ children }: { children: React.ReactNode }) {
  return <div data-testid="container">{children}</div>;
}

function customRender(
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
): RenderResult {
  return render(ui, { wrapper: AllTheProviders, ...options });
}

function firstComponent(screen: Screen) {
  return screen.getByTestId("container").firstChild;
}

export * from "@testing-library/react";
export * from "@testing-library/user-event";
export { customRender as render, firstComponent };
import "@testing-library/jest-dom";
import { usePathname } from "next/navigation";

jest.mock("next/navigation", () => {
  return {
    useRouter() {
      return { push: () => {} };
    },
    useParams: () => ({}),
    notFound: () => null,
    usePathname: () => null,
    useSearchParams: () => null,
  };
});
