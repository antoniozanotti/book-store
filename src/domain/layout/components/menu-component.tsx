"use client";
import { UiCollapsibleComponent } from "@/ui-collapsible-component";
import { MenuItemComponent } from "./menu-item-component";
import { useMenuContext } from "../contexts/menu-context";
import { useLayoutContext } from "@/domain/layout/contexts/layout-context";
import { getCategoryUrl } from "@/domain/category/helpers/get-category-url";
import { Divider } from "@nextui-org/react";
import { Fragment } from "react";

export function MenuComponent({ ...props }: React.ComponentProps<"nav">) {
  const menuState = useMenuContext().state;
  const showMenuMobileStyles = menuState.isOpen ? "" : "hidden";
  const layoutState = useLayoutContext().state;
  const categories = layoutState.categories;
  const navStyles = `${showMenuMobileStyles} lg:block`;
  props.className = props.className
    ? `${navStyles} ${props.className}`
    : navStyles;

  return (
    <nav {...props} aria-label="Main Navigation">
      <ul className="flex flex-col lg:flex-row justify-center lg:w-fit mx-auto overflow-hidden lg:gap-[2px] lg:rounded-full leading-[16px] text-primary-800">
        <MenuItemComponent href="/about-us" title="About Us" />
        <FormatedDivider />
        <MenuItemComponent>
          <UiCollapsibleComponent
            theHeader="Books"
            theContent={
              <ul className="text-sm">
                {categories.map((category, index) => (
                  <Fragment key={category.id}>
                    <MenuItemComponent
                      href={getCategoryUrl(category)}
                      title={category.name}
                      className="bg-primary-500 hover:bg-primary-400"
                    />
                    {index < categories.length - 1 && <FormatedDivider />}
                  </Fragment>
                ))}
              </ul>
            }
          />
        </MenuItemComponent>
        <FormatedDivider />
        <MenuItemComponent href="/contact-us" title="Contact Us" />
        <FormatedDivider />
        <MenuItemComponent href="/blog" title="Blog" />
      </ul>
    </nav>
  );
}

function FormatedDivider() {
  return <Divider className="bg-primary-200 lg:hidden" />;
}