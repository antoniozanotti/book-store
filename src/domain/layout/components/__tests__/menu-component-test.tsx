import { render, screen } from "@/lib/test";
import { MenuComponent } from "../menu-component";
import { LayoutProvider } from "../../contexts/layout-context";
import { categoriesWithPageMock } from "@/domain/category/__tests__/categories-with-page-mock";

describe("MenuComponent", () => {
  it("should render", async () => {
    const { container } = render(
      <LayoutProvider categories={categoriesWithPageMock}>
        <MenuComponent />
      </LayoutProvider>
    );
    expect(container).toMatchSnapshot();
  });
});
