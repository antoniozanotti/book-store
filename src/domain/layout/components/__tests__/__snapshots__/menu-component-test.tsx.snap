// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuComponent should render 1`] = `
<div>
  <div
    data-testid="container"
  >
    <nav
      aria-label="Main Navigation"
      class="hidden lg:block"
    >
      <ul
        class="flex flex-col lg:flex-row justify-center lg:w-fit mx-auto overflow-hidden lg:gap-[2px] lg:rounded-full leading-[16px] text-primary-800"
      >
        <li
          class="bg-primary-300 hover:bg-primary-400"
        >
          <a
            class="p-4 lg:px-10 lg:py-2 block"
            href="/about-us"
            title="About Us"
          >
            About Us
          </a>
        </li>
        <hr
          class="shrink-0 border-none w-full h-divider bg-primary-200 lg:hidden"
          role="separator"
        />
        <li
          class="bg-primary-300 hover:bg-primary-400"
        >
          <div>
            <button
              aria-expanded="false"
              aria-haspopup="true"
              class="flex cursor-pointer gap-2 p-4 lg:px-10 lg:py-2 items-center"
              tabindex="0"
            >
              <span
                class="grow"
              >
                Books
              </span>
              <svg
                class="lucide lucide-chevron-left size-4"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m15 18-6-6 6-6"
                />
              </svg>
            </button>
            <div
              class="hidden"
            >
              <ul
                class="text-sm"
              >
                <li
                  class="bg-primary-300 hover:bg-primary-400 bg-primary-500 hover:bg-primary-400"
                >
                  <a
                    class="p-4 lg:px-10 lg:py-2 block"
                    href="/category/arts-and-entertainment"
                    title="Arts and Entertainment"
                  >
                    Arts and Entertainment
                  </a>
                </li>
                <hr
                  class="shrink-0 border-none w-full h-divider bg-primary-200 lg:hidden"
                  role="separator"
                />
                <li
                  class="bg-primary-300 hover:bg-primary-400 bg-primary-500 hover:bg-primary-400"
                >
                  <a
                    class="p-4 lg:px-10 lg:py-2 block"
                    href="/category/business-and-technology"
                    title="Business and Technology"
                  >
                    Business and Technology
                  </a>
                </li>
                <hr
                  class="shrink-0 border-none w-full h-divider bg-primary-200 lg:hidden"
                  role="separator"
                />
                <li
                  class="bg-primary-300 hover:bg-primary-400 bg-primary-500 hover:bg-primary-400"
                >
                  <a
                    class="p-4 lg:px-10 lg:py-2 block"
                    href="/category/graphic-novels"
                    title="Graphic Novels"
                  >
                    Graphic Novels
                  </a>
                </li>
                <hr
                  class="shrink-0 border-none w-full h-divider bg-primary-200 lg:hidden"
                  role="separator"
                />
                <li
                  class="bg-primary-300 hover:bg-primary-400 bg-primary-500 hover:bg-primary-400"
                >
                  <a
                    class="p-4 lg:px-10 lg:py-2 block"
                    href="/category/history-biography-and-social-science"
                    title="History, Biography, and Social Science"
                  >
                    History, Biography, and Social Science
                  </a>
                </li>
                <hr
                  class="shrink-0 border-none w-full h-divider bg-primary-200 lg:hidden"
                  role="separator"
                />
                <li
                  class="bg-primary-300 hover:bg-primary-400 bg-primary-500 hover:bg-primary-400"
                >
                  <a
                    class="p-4 lg:px-10 lg:py-2 block"
                    href="/category/home-and-lifestyle"
                    title="Home and Lifestyle"
                  >
                    Home and Lifestyle
                  </a>
                </li>
                <hr
                  class="shrink-0 border-none w-full h-divider bg-primary-200 lg:hidden"
                  role="separator"
                />
                <li
                  class="bg-primary-300 hover:bg-primary-400 bg-primary-500 hover:bg-primary-400"
                >
                  <a
                    class="p-4 lg:px-10 lg:py-2 block"
                    href="/category/horror"
                    title="Horror"
                  >
                    Horror
                  </a>
                </li>
                <hr
                  class="shrink-0 border-none w-full h-divider bg-primary-200 lg:hidden"
                  role="separator"
                />
                <li
                  class="bg-primary-300 hover:bg-primary-400 bg-primary-500 hover:bg-primary-400"
                >
                  <a
                    class="p-4 lg:px-10 lg:py-2 block"
                    href="/category/languages-and-travel"
                    title="Languages and Travel"
                  >
                    Languages and Travel
                  </a>
                </li>
                <hr
                  class="shrink-0 border-none w-full h-divider bg-primary-200 lg:hidden"
                  role="separator"
                />
                <li
                  class="bg-primary-300 hover:bg-primary-400 bg-primary-500 hover:bg-primary-400"
                >
                  <a
                    class="p-4 lg:px-10 lg:py-2 block"
                    href="/category/lgbtqia"
                    title="LGBTQIA+"
                  >
                    LGBTQIA+
                  </a>
                </li>
                <hr
                  class="shrink-0 border-none w-full h-divider bg-primary-200 lg:hidden"
                  role="separator"
                />
                <li
                  class="bg-primary-300 hover:bg-primary-400 bg-primary-500 hover:bg-primary-400"
                >
                  <a
                    class="p-4 lg:px-10 lg:py-2 block"
                    href="/category/literature-and-poetry"
                    title="Literature and Poetry"
                  >
                    Literature and Poetry
                  </a>
                </li>
                <hr
                  class="shrink-0 border-none w-full h-divider bg-primary-200 lg:hidden"
                  role="separator"
                />
                <li
                  class="bg-primary-300 hover:bg-primary-400 bg-primary-500 hover:bg-primary-400"
                >
                  <a
                    class="p-4 lg:px-10 lg:py-2 block"
                    href="/category/mystery"
                    title="Mystery"
                  >
                    Mystery
                  </a>
                </li>
                <hr
                  class="shrink-0 border-none w-full h-divider bg-primary-200 lg:hidden"
                  role="separator"
                />
                <li
                  class="bg-primary-300 hover:bg-primary-400 bg-primary-500 hover:bg-primary-400"
                >
                  <a
                    class="p-4 lg:px-10 lg:py-2 block"
                    href="/category/romance"
                    title="Romance"
                  >
                    Romance
                  </a>
                </li>
                <hr
                  class="shrink-0 border-none w-full h-divider bg-primary-200 lg:hidden"
                  role="separator"
                />
                <li
                  class="bg-primary-300 hover:bg-primary-400 bg-primary-500 hover:bg-primary-400"
                >
                  <a
                    class="p-4 lg:px-10 lg:py-2 block"
                    href="/category/science-and-math"
                    title="Science and Math"
                  >
                    Science and Math
                  </a>
                </li>
                <hr
                  class="shrink-0 border-none w-full h-divider bg-primary-200 lg:hidden"
                  role="separator"
                />
                <li
                  class="bg-primary-300 hover:bg-primary-400 bg-primary-500 hover:bg-primary-400"
                >
                  <a
                    class="p-4 lg:px-10 lg:py-2 block"
                    href="/category/science-fiction-and-fantasy"
                    title="Science Fiction and Fantasy"
                  >
                    Science Fiction and Fantasy
                  </a>
                </li>
                <hr
                  class="shrink-0 border-none w-full h-divider bg-primary-200 lg:hidden"
                  role="separator"
                />
                <li
                  class="bg-primary-300 hover:bg-primary-400 bg-primary-500 hover:bg-primary-400"
                >
                  <a
                    class="p-4 lg:px-10 lg:py-2 block"
                    href="/category/spirituality-and-philosophy"
                    title="Spirituality and Philosophy"
                  >
                    Spirituality and Philosophy
                  </a>
                </li>
                <hr
                  class="shrink-0 border-none w-full h-divider bg-primary-200 lg:hidden"
                  role="separator"
                />
                <li
                  class="bg-primary-300 hover:bg-primary-400 bg-primary-500 hover:bg-primary-400"
                >
                  <a
                    class="p-4 lg:px-10 lg:py-2 block"
                    href="/category/wellness-sports-and-outdoors"
                    title="Wellness, Sports, and Outdoors"
                  >
                    Wellness, Sports, and Outdoors
                  </a>
                </li>
                <hr
                  class="shrink-0 border-none w-full h-divider bg-primary-200 lg:hidden"
                  role="separator"
                />
                <li
                  class="bg-primary-300 hover:bg-primary-400 bg-primary-500 hover:bg-primary-400"
                >
                  <a
                    class="p-4 lg:px-10 lg:py-2 block"
                    href="/category/young-adult"
                    title="Young Adult"
                  >
                    Young Adult
                  </a>
                </li>
              </ul>
            </div>
          </div>
        </li>
        <hr
          class="shrink-0 border-none w-full h-divider bg-primary-200 lg:hidden"
          role="separator"
        />
        <li
          class="bg-primary-300 hover:bg-primary-400"
        >
          <a
            class="p-4 lg:px-10 lg:py-2 block"
            href="/contact-us"
            title="Contact Us"
          >
            Contact Us
          </a>
        </li>
        <hr
          class="shrink-0 border-none w-full h-divider bg-primary-200 lg:hidden"
          role="separator"
        />
        <li
          class="bg-primary-300 hover:bg-primary-400"
        >
          <a
            class="p-4 lg:px-10 lg:py-2 block"
            href="/blog"
            title="Blog"
          >
            Blog
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;
